frontend/src/components/EmbeddedApp.tsx

```diff
@@ -0,0 +1,201 @@
+import {
+  AppIFrameMessage,
+  AppIFrameRequest,
+  AppIFrameResponse,
+} from "@/types/app-iframe-message";
+import styled from "@emotion/styled";
+import {
+  forwardRef,
+  useCallback,
+  useEffect,
+  useImperativeHandle,
+  useRef,
+  useState,
+} from "react";
+
+const StyledIFrame = styled.iframe`
+  width: 100%;
+  border: none;
+
+  flex-grow: 1;
+`;
+
+const MAX_COUNTER = 1000000;
+let counter = 0;
+
+const postMessageToIFrame = (
+  iFrame: HTMLIFrameElement,
+  message: AppIFrameMessage,
+) => {
+  if (!iFrame.contentWindow) {
+    console.error(
+      "Cannot post message to iframe without content window:",
+      iFrame,
+    );
+    return;
+  }
+
+  // get target origin from the iframe's src attribute
+  const targetOrigin = new URL(iFrame.src).origin;
+
+  iFrame.contentWindow.postMessage(message, {
+    targetOrigin,
+  });
+};
+
+export interface EmbeddedAppRef {
+  sendRequest<ProcedureName extends AppIFrameRequest["procedure"]>(
+    request: Omit<AppIFrameRequest, "id" | "type"> & {
+      procedure: ProcedureName;
+    },
+  ): Promise<AppIFrameResponse & { procedure: ProcedureName }>;
+}
+
+export interface Props {
+  src: string;
+}
+
+const EmbeddedApp = forwardRef<EmbeddedAppRef, Props>(function EmbeddedApp(
+  { src },
+  ref,
+) {
+  const [iFrame, setIFrame] = useState<HTMLIFrameElement | null>(null);
+  const isIFrameLoaded = useRef<boolean>(false);
+
+  const pendingRequests = useRef<{
+    [key: number]: (response: AppIFrameResponse) => void;
+  }>({});
+
+  const sendRequest = useCallback(
+    <ProcedureName extends AppIFrameMessage["procedure"]>(
+      request: Omit<AppIFrameRequest, "id" | "type"> & {
+        procedure: ProcedureName;
+      },
+    ): Promise<AppIFrameResponse & { procedure: ProcedureName }> => {
+      if (iFrame && isIFrameLoaded.current) {
+        return new Promise((resolve) => {
+          // store the resolve function in the pendingRequests object
+          pendingRequests.current[counter] = (response: AppIFrameResponse) => {
+            if (response.procedure !== request.procedure) {
+              console.error("Invalid response procedure", response, request);
+              return;
+            }
```

Comment:
if we're already generating unique IDs and returning them in the response, we don't really need to return the `procedure` as well, do we? 🤔

Reply:
Yes you are right although it makes it more difficult from a typing perspective to differentiate the different responses..

Reply:
Yes, but we'd need type narrowing anyways, since the other app may not be fully trusted.

Reply:
yes, true!